src/containers/multiple/multiple.container.tsx

```diff
@@ -0,0 +1,93 @@
+import {Component, Description, Label} from '@components';
+import {IRenderable, RenderComponent} from '@lib/renderer';
+import {IComponentProps, Values} from '@types';
+import {ComponentSchema} from 'formiojs';
+import React, {useState} from 'react';
+
+export interface IMultipleComponent extends ComponentSchema {
+  type: string;
+}
+
+export interface IMultipleProps extends IComponentProps {
+  component: IMultipleComponent;
+  value: Values;
+}
+
+/**
+ * Implements `multiple: true` behaviour.
+ *
+ * Provide a thin wrapper around a component with controls for adding multiple instances. Utilizes
+ * <RenderComponent/> to render individual instances.
+ */
+export const Multiple: React.FC<IMultipleProps> = props => {
+  const {component, form, path, value = [], setValue} = props; // FIXME: Awaits future pr.
+  const [keys, setKeys] = useState<number[]>(Array.from({length: 1}, (_, i) => i));
```

Comment:
Yes this can be fixed, At first I had a separate count state which was then used here to generate the keys. I forgot to remove this after normalizing.

> Will all returned components then have a reference to the same Array instance, the way Python default parameter values shouldn't be a mutable?

All rendered components should receive a clone of the object definition. The keys state here has noting to do with that. It's a mechanism to provide stable id's. Since we clone the component, we have no reliable unique data, we can't rely on index because:

Given: 3 items are rendered: ['lorem', 'ipsum', 'dolor'] with indexes as keys.
Then 'ipsum' will have key 1.
When I remove 'ipsum'.
Then the next render cycle will have 'dolor'  as key 1. This causes React to handle 'dolor' as 'ipsum'.

This mechanism/hack provides a workaround:

Given: 3 items are rendered: ['lorem', 'ipsum', 'dolor'] with state [0, 1, 2] as keys.
Then 'ipsum' will have key 1.
When I remove 'ipsum'.
And thereby remove key 1 from state.
Then the next render cycle will have 'dolor'  will still have key 2.
When I then add another component.
And I generate a new key for based on the highest key in use.
Then it will be rendered with 3 as key.

Reply:
None of this mechanism/hack is required when you use the formik FieldArray component, fyi. It provides handlers and allows you to map over the values in rendering. Removal, addition and insertion work correctly there.

Reply:
I know, but as discussed earlier: I don’t want to rely on implementation details outside the renderer itself (and rely solely on the provided interface) as this would be problematic with third party implementations.